go/libraries/events/file_backed_proc.go

```diff
@@ -0,0 +1,235 @@
+
+// fileChecingFunc is the signature used for functions used to authenticate
+// filenames created by the fileNamingFunc
+type fileCheckingFunc func(data []byte, path string) (bool, error)
+
+// MD5Str returns the standard base64 encoding of the md5 hash with padding
+func MD5Str(bytes []byte) string {
+	md5Bytes := md5.Sum(bytes)
+	str := base64.StdEncoding.EncodeToString(md5Bytes[:])
+	return str
+}
+
+// MD5StrUrl returns the base64 url encoding of the md5 hash with the padding removed
+func MD5StrUrl(bytes []byte) string {
+	md5Bytes := md5.Sum(bytes)
+	str := base64.URLEncoding.EncodeToString(md5Bytes[:])
+	return str[:22]
+}
+
+// MD5FileNamer names files after the base64 url encoding of the md5 hash of the contents of the file
+func MD5FileNamer(bytes []byte) string {
+	return MD5StrUrl(bytes) + evtDataExt
+}
+
+// CheckFilenameMD5 is the fileNamingFunc used when instantiating a FileBackedProc
+func CheckFilenameMD5(data []byte, path string) (bool, error) {
+	filename := filepath.Base(path)
+	ext := filepath.Ext(filename)
+
+	if ext != evtDataExt {
+		return false, nil
+	}
+
+	md5FromFilename := filename[:len(filename)-len(ext)]
+	md5FromData := MD5StrUrl(data)
+
+	if md5FromFilename != md5FromData {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+var errEventsDirNotExists = errors.New("no events data directory exists")
+
+// eventsDataDir is the directory used to store the events requests files
+type eventsDataDir struct {
+	fs   filesys.Filesys
+	path string
+}
+
+// newEventsDataDir creates a new eventsDataDir
+func newEventsDataDir(fs filesys.Filesys, homeDir string, doltDir string) *eventsDataDir {
+	path := filepath.Join(homeDir, doltDir, eventsDir)
+
+	return &eventsDataDir{fs: fs, path: path}
+}
+
+//  MakeEventsDir creates a new events data dir in the main dolt dir
+func (evd *eventsDataDir) MakeEventsDir() error {
+	if exists, _ := evd.fs.Exists(evd.path); !exists {
+		if err := evd.fs.MkDirs(evd.path); err != nil {
+			return err
+		}
+		return nil
+	}
+	return nil
+}
+
+func (evd *eventsDataDir) getPath() string {
+	return evd.path
+}
+
+// FileBackedProc writes events requests to files in an events data dir
+type FileBackedProc struct {
+	ed           *eventsDataDir
+	namingFunc   fileNamingFunc
+	CheckingFunc fileCheckingFunc
+	LockPath     string
+}
+
+// NewFileBackedProc creates a new FileBackedProc
+func NewFileBackedProc(fs filesys.Filesys, userHomeDir string, doltDir string, nf fileNamingFunc, cf fileCheckingFunc) *FileBackedProc {
+	eventsDataDir := newEventsDataDir(fs, userHomeDir, doltDir)
+
+	if err := eventsDataDir.MakeEventsDir(); err != nil {
+		panic(err)
+	}
+
+	lp := filepath.Join(eventsDataDir.getPath(), doltLockFile)
+
+	exists, _ := fs.Exists(lp)
+
+	if !exists {
+		if err := fs.WriteFile(lp, []byte("lockfile for dolt \n")); err != nil {
+			panic(err)
+		}
+	}
+
+	return &FileBackedProc{ed: eventsDataDir, namingFunc: nf, CheckingFunc: cf, LockPath: lp}
+}
+
```

Comment:
This is the new implementation of the locking system. Instead of locking the entire events directory, I create a `dolt.lock` file if it does not exist, then lock/unlock this file.